File: apps/web/src/lib/api/job-register.ts
import { apiRequest } from "@/lib/api-client";
import type { Page } from "@/lib/api/types";

export type JobWorkStatus = "not_completed" | "in_progress" | "completed";
export type JobTaxInvoiceStatus = "not_submitted" | "submitted";
export type JobPaymentStatus = "lpo_not_received" | "pending" | "received" | "not_received";

export interface JobRegisterRow {
  id: string;
  qt_no: string | null;
  quotation_id: string | null;
  lpo_no: string | null;
  villa_no: string | null;
  description: string;
  rate: string;
  vat: string;
  override_total: string | null;
  total: string;
  customer_id: string | null;
  company_text: string | null;
  work_status: JobWorkStatus;
  tax_invoice_status: JobTaxInvoiceStatus;
  payment_status: JobPaymentStatus;
  remarks: string | null;
  source_invoice_id: string | null;
  sort_order: number;
}

export interface JobRegisterRowInput {
  qt_no?: string | null;
  quotation_id?: string | null;
  lpo_no?: string | null;
  villa_no?: string | null;
  description: string;
  rate?: string;
  override_total?: string | null;
  customer_id?: string | null;
  company_text?: string | null;
  work_status?: JobWorkStatus;
  tax_invoice_status?: JobTaxInvoiceStatus;
  payment_status?: JobPaymentStatus;
  remarks?: string | null;
  source_invoice_id?: string | null;
  sort_order?: number;
}

export interface JobRegisterRowUpdate extends Partial<JobRegisterRowInput> {
  clear_override_total?: boolean;
}

export interface JobReceipt {
  id: string;
  customer_id: string | null;
  company_text: string | null;
  amount: string;
  receipt_date: string;
  note: string | null;
  linked_payment_id: string | null;
}

export interface JobReceiptInput {
  customer_id?: string | null;
  company_text?: string | null;
  amount: string;
  receipt_date: string;
  note?: string | null;
  linked_payment_id?: string | null;
}

export interface JobRegisterSummary {
  total_work_value: string;
  total_submitted: string;
  total_received: string;
  total_received_by_status: string;
  balance: string;
}

export interface JobRegisterListParams {
  customer_id?: string;
  work_status?: JobWorkStatus;
  tax_invoice_status?: JobTaxInvoiceStatus;
  payment_status?: JobPaymentStatus;
  date_from?: string;
  date_to?: string;
  search?: string;
  page?: number;
  pageSize?: number;
}

function buildQuery(params: JobRegisterListParams): string {
  const query = new URLSearchParams();
  if (params.customer_id) query.set("customer_id", params.customer_id);
  if (params.work_status) query.set("work_status", params.work_status);
  if (params.tax_invoice_status) query.set("tax_invoice_status", params.tax_invoice_status);
  if (params.payment_status) query.set("payment_status", params.payment_status);
  if (params.date_from) query.set("date_from", params.date_from);
  if (params.date_to) query.set("date_to", params.date_to);
  if (params.search) query.set("search", params.search);
  if (params.page) query.set("page", String(params.page));
  if (params.pageSize) query.set("page_size", String(params.pageSize));
  const qs = query.toString();
  return qs ? `?${qs}` : "";
}

export const jobRegisterApi = {
  list: (params: JobRegisterListParams = {}) =>
    apiRequest<Page<JobRegisterRow>>(`/job-register/rows${buildQuery(params)}`),
  summary: (params: JobRegisterListParams = {}) =>
    apiRequest<JobRegisterSummary>(`/job-register/summary${buildQuery(params)}`),
  createRow: (payload: JobRegisterRowInput) =>
    apiRequest<JobRegisterRow>("/job-register/rows", { method: "POST", body: JSON.stringify(payload) }),
  updateRow: (id: string, payload: JobRegisterRowUpdate) =>
    apiRequest<JobRegisterRow>(`/job-register/rows/${id}`, { method: "PATCH", body: JSON.stringify(payload) }),
  removeRow: (id: string) => apiRequest<void>(`/job-register/rows/${id}`, { method: "DELETE" }),
  listReceipts: () => apiRequest<JobReceipt[]>("/job-register/receipts"),
  createReceipt: (payload: JobReceiptInput) =>
    apiRequest<JobReceipt>("/job-register/receipts", { method: "POST", body: JSON.stringify(payload) }),
  updateReceipt: (id: string, payload: Partial<JobReceiptInput>) =>
    apiRequest<JobReceipt>(`/job-register/receipts/${id}`, { method: "PATCH", body: JSON.stringify(payload) }),
  removeReceipt: (id: string) => apiRequest<void>(`/job-register/receipts/${id}`, { method: "DELETE" }),
};

File: apps/api/app/schemas/job_register.py
import uuid
from datetime import date
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field, computed_field

from app.models.enums import JobPaymentStatus, JobTaxInvoiceStatus, JobWorkStatus


class JobRegisterRowCreate(BaseModel):
    qt_no: str | None = Field(default=None, max_length=100)
    quotation_id: uuid.UUID | None = None
    lpo_no: str | None = Field(default=None, max_length=100)
    villa_no: str | None = Field(default=None, max_length=100)
    description: str = Field(min_length=1, max_length=500)
    rate: Decimal = Decimal("0")
    override_total: Decimal | None = None
    customer_id: uuid.UUID | None = None
    company_text: str | None = Field(default=None, max_length=255)
    work_status: JobWorkStatus = JobWorkStatus.NOT_COMPLETED
    tax_invoice_status: JobTaxInvoiceStatus = JobTaxInvoiceStatus.NOT_SUBMITTED
    payment_status: JobPaymentStatus = JobPaymentStatus.PENDING
    remarks: str | None = Field(default=None, max_length=1000)
    source_invoice_id: uuid.UUID | None = None
    sort_order: int = 0
    client_uuid: uuid.UUID | None = None


class JobRegisterRowUpdate(BaseModel):
    qt_no: str | None = Field(default=None, max_length=100)
    quotation_id: uuid.UUID | None = None
    lpo_no: str | None = Field(default=None, max_length=100)
    villa_no: str | None = Field(default=None, max_length=100)
    description: str | None = Field(default=None, min_length=1, max_length=500)
    rate: Decimal | None = None
    override_total: Decimal | None = None
    clear_override_total: bool = False
    customer_id: uuid.UUID | None = None
    company_text: str | None = Field(default=None, max_length=255)
    work_status: JobWorkStatus | None = None
    tax_invoice_status: JobTaxInvoiceStatus | None = None
    payment_status: JobPaymentStatus | None = None
    remarks: str | None = Field(default=None, max_length=1000)
    source_invoice_id: uuid.UUID | None = None
    sort_order: int | None = None


class JobRegisterRowOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    qt_no: str | None
    quotation_id: uuid.UUID | None
    lpo_no: str | None
    villa_no: str | None
    description: str
    rate: Decimal
    vat: Decimal
    override_total: Decimal | None
    customer_id: uuid.UUID | None
    company_text: str | None
    work_status: JobWorkStatus
    tax_invoice_status: JobTaxInvoiceStatus
    payment_status: JobPaymentStatus
    remarks: str | None
    source_invoice_id: uuid.UUID | None
    sort_order: int

    @computed_field  # type: ignore[prop-decorator]
    @property
    def total(self) -> Decimal:
        return self.override_total if self.override_total is not None else self.rate + self.vat


class JobReceiptCreate(BaseModel):
    customer_id: uuid.UUID | None = None
    company_text: str | None = Field(default=None, max_length=255)
    amount: Decimal = Field(gt=0)
    receipt_date: date
    note: str | None = Field(default=None, max_length=1000)
    linked_payment_id: uuid.UUID | None = None
    client_uuid: uuid.UUID | None = None


class JobReceiptUpdate(BaseModel):
    customer_id: uuid.UUID | None = None
    company_text: str | None = Field(default=None, max_length=255)
    amount: Decimal | None = Field(default=None, gt=0)
    receipt_date: date | None = None
    note: str | None = Field(default=None, max_length=1000)
    linked_payment_id: uuid.UUID | None = None


class JobReceiptOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    customer_id: uuid.UUID | None
    company_text: str | None
    amount: Decimal
    receipt_date: date
    note: str | None
    linked_payment_id: uuid.UUID | None


class JobRegisterSummary(BaseModel):
    total_work_value: Decimal
    total_submitted: Decimal
    total_received: Decimal
    total_received_by_status: Decimal
    balance: Decimal
